### ISSUE TITLE
non-fork 'hunterization' with patchfile

### ISSUE THREAD

**Issue by cinterloper** [ruslo/hunter#1528]
Im a little uncomfortable with 'forking' packages just to add 'hunterization'
Might it be more reasonable to maintain patchfiles in the hunter tree, and use PATCH_COMMAND to apply them instead of maintaining a fork?

an example of what im describing:

https://github.com/digego/extempore/blob/master/CMakeLists.txt#L316

I believe 
https://github.com/hunter-packages/clang/commit/24c45ce2b557b18715456c9b8aae5742d12d2dc5

is a simple change, that would be encapsulated by using PATCH_COMMAND inside the hunter_add_version(...) declaration alongside the patch file in the main hunter/$PROJECT source tree

```
--- a/tools/CMakeLists.txt
+++ b/tools/CMakeLists.txt
@@ -20,4 +20,15 @@ endif()
 # on top of the Clang tooling platform. We keep them in a separate repository
 # to keep the primary Clang repository small and focused.
 # It also may be included by LLVM_EXTERNAL_CLANG_TOOLS_EXTRA_SOURCE_DIR.
+if(HUNTER_ENABLED)
+  hunter_add_package(ClangToolsExtra) # CLANGTOOLSEXTRA_ROOT
+  set(
+      LLVM_EXTERNAL_CLANG_TOOLS_EXTRA_SOURCE_DIR
+      "${CLANGTOOLSEXTRA_ROOT}"
+      CACHE
+      PATH
+      "Path to Clang tools extra"
+      FORCE
+  )
+endif()
 add_llvm_external_project(clang-tools-extra extra)
--

```

**Comment by xsacha** [comment]
Makes it easy to track the changes, apply changes to future versions and also do pull requests upstream.
The patches would still need to be versioned (such as myproject v0.5-p1) as they would likely differ between commits.

**Comment by ruslo** [comment]
The logic is next: if the patch is trivial then maintaining a fork is a trivial effort, also you can just send the patch upstream reducing maintenance to zero. If the patch is not trivial then supporting set of `*.patch` files is not practical - fork is more clear solution: https://docs.hunter.sh/en/latest/faq/why-do-we-need-forks.html#set-of-patch-files 

> an example of what im describing:
> https://github.com/digego/extempore/blob/master/CMakeLists.txt#L316

Currently cache functionality relies on fact that sources = SHA1 of archive, if you add patches you have to design the system of applying SHA1 of patches.

**Comment by tonyrewin** [comment]
It looks pretty much like a bottleneck here. Maybe caching and hunterization logics should be more isolated and play more like the other package managers with an ability to pick a version and to use a custom repository address.

**Comment by ruslo** [comment]
> Maybe caching and hunterization logics should be more isolated and play more like the other package managers with an ability to pick a version and to use a custom repository address.

How it is related to keeping tons of patches in Hunter repository instead of keeping them separately in place where they logically belongs to?

About "ability to pick a version":
* https://docs.hunter.sh/en/latest/overview/customization/config-id.html
* https://docs.hunter.sh/en/latest/reference/user-modules/hunter_config.html

About "custom repository address":
* https://docs.hunter.sh/en/latest/reference/user-variables.html#hunter-download-server

**Comment by mchiasson** [comment]
@ruslo I have a good use-case for supporting patches:  PlayStation®4 - SCE DevNet

Sony has a strict non-disclosure agreement that doesn't allow us to to expose anything from their SDK publicly. We have encountered a few third party libraries that had to be modified for PS4 use, and we had no choice but to make private forks of these third party libraries. 

If we were able to provide private patch files through the `hunter_config` command in our `cmake/Hunter/config.cmake`, it would remove the need of doing private forks. 

Personally, I prefer private forks as they are easier to maintain than maintaining patch files when merge conflicts occur, as long as my employer pays for those private forks 😅
